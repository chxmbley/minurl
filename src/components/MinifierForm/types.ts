import { URL_FIELD_NAME } from '~components/MinifierForm/constants';

/** Form fields and data types for MinifierForm */
export type MinifierFormData = {
  [URL_FIELD_NAME]: string;
};

/** Props for the MinifierForm component */
export type MinifierFormProps = {
  /** CSS class to apply to the component */
  className?: string;
  /** Whether to render a loading state and disable user input */
  isLoading?: boolean;
  /** Callback executed when the form is reset */
  onReset: () => void;
  /** Callback executed with form data when the form is submitted */
  onSubmit: (data: MinifierFormData) => void;
  /** Shortened URL slug to apply to the form field for the user to copy. If null, the field is editable. */
  slug: string | null;
};
