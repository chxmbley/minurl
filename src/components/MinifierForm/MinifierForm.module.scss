.urlInput {
  @apply w-full pr-12;
}

.submitButton {
  @apply absolute right-1 top-1 w-10 h-10 border border-transparent;

  &:disabled {
    @apply border-slate-200 dark:border-slate-700;
  }
}

.submitButtonIcon {
  @apply w-5 h-5 flex-shrink-0;
}

.copyButton {
  @apply absolute right-12 top-1/2 -translate-y-1/2 w-20;

  &.copied {
    @apply text-green-600 dark:text-green-400;

    &:hover,
    &:focus {
      @apply hover:text-green-600 hover:dark:text-green-300;
    }
  }
}
