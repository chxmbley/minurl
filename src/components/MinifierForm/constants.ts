/** Name of the URL input field in the MinifierForm form data */
export const URL_FIELD_NAME = 'url';
