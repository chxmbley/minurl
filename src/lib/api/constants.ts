export const MINIFY_ENDPOINT = '/api/v1/minify';
