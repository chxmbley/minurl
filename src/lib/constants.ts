/**
 * Base URL of the application. This is the URL to which shortened URL slugs will be appended.
 * The base URL will be read from the NEXT_PUBLIC_BASE_URL if available and fallback to the app URL
 * as determined by the Location object (client rendered only)
 */
export const APP_BASE_URL = (() => {
  const locationUrl = typeof window === 'undefined' ? '' : `${location.protocol}//${location.host}`;
  return process.env.NEXT_PUBLIC_BASE_URL ?? locationUrl;
})();

/**
 * List of display messages for short input URLs
 */
export const SHORT_INPUT_URL_MESSAGES = ["that's not *too* long...", 'hmm... already seems short to me', 'v smol url'];

/**
 * List of display messages for long input URLs
 */
export const LONG_INPUT_URL_MESSAGES = [
  'wow many large. such url.',
  'is that a real url or did a cat walk across your keyboard?',
  'thicc url alert',
  'more like www.toomanyletters.com amirite?',
  "back in my day, urls weren't a mile long",
];

/**
 * List of display messages for output URLs
 */
export const OUTPUT_URL_MESSAGES = [
  'much better',
  'sweet relief',
  'reject verbosity. embrace brevity.',
  'with mini urls comes great responsibility',
  'little url â€” big dreams',
  '*chef kiss*',
];
