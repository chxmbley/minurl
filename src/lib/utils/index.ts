import { createHash } from 'crypto';

/**
 * Creates an MD5 hash (hexadecimal) of the provided string
 * @param str - String to hash
 */
export const md5Hash = (str: string): string => createHash('md5').update(str).digest('hex');

/**
 * Randomly selects one element from a given array
 * @param arr - Array from which to select the element
 */
export const getRandomElement = <T>(arr: T[]): T => arr[Math.floor(Math.random() * arr.length)];
