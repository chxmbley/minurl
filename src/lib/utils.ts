import { createHash } from 'crypto';
import RandExp from 'randexp';
import { APP_BASE_URL, LONG_INPUT_URL_MESSAGES, OUTPUT_URL_MESSAGES, SHORT_INPUT_URL_MESSAGES } from './constants';

/**
 * Generates a random alphanumeric slug.
 * @param length - Number of characters in the resulting string. Default value is 6.
 * Throws an error if the length is less than 3 or greater than 32.
 */
export const generateRandomSlug = (length = 6): string => {
  if (length < 3 || length > 32) {
    throw new Error('Length must be between 3 and 32');
  }

  const slugRegex = new RegExp(`[a-z0-9]{${length}}`, 'i');
  return new RandExp(slugRegex).gen();
};

/**
 * Returns a boolean whether the provided URL is valid. URL must begin with http(s) protocol.
 * @param url - URL to test
 */
export const isValidUrl = (url: string): boolean => {
  try {
    new URL(url);
  } catch (_) {
    return false;
  }

  return /^http(s)?:\/\//i.test(url);
};

/**
 * Constructs a minified URL using the configured base URL and provided slug
 * @param slug - The URL slug to append to the base URL
 */
export const constructMiniUrlFromSlug = (slug: string): string => `${APP_BASE_URL}/${slug}`;

/**
 * Checks whether a provided URL is an app redirect URL. That is, a URL that may be used by this
 * application to host a redirect. Note that this function does not check whether the URL exists.
 * @param url - URL to test
 */
export const isAppRedirectUrl = (url: string): boolean => {
  try {
    const { host, pathname } = new URL(url);
    return host === new URL(APP_BASE_URL).host && /^\/[a-z0-9]+$/i.test(pathname);
  } catch (_) {
    return false;
  }
};

/**
 * Randomly selects one element from a given array
 * @param arr - Array from which to select the element
 */
const getRandomElement = <T>(arr: T[]): T => arr[Math.floor(Math.random() * arr.length)];

/**
 * Selects a random message to display for an input URL.
 * The length of the provided URL determines which list of messages the resulting message is selected from.
 * @param url - Input URL for which to generate a message
 */
export const getRandomInputUrlMessage = (url: string): string => {
  const messageList = url.length < 45 ? SHORT_INPUT_URL_MESSAGES : LONG_INPUT_URL_MESSAGES;
  return getRandomElement(messageList);
};

/**
 * Selects a random message to display for an output (shortened) URL.
 */
export const getRandomOutputUrlMessage = (): string => getRandomElement(OUTPUT_URL_MESSAGES);

/**
 * Creates an MD5 hash (hexadecimal) of the provided string
 * @param str - String to hash
 */
export const md5Hash = (str: string): string => createHash('md5').update(str).digest('hex');
