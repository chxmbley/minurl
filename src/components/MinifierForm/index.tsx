import cn from 'classnames';
import { isEmpty } from 'lodash';
import { useCallback, useEffect, useMemo, useState } from 'react';
import { ArrowRightIcon, RewindIcon } from '@heroicons/react/solid';
import { useForm } from 'react-hook-form';
import Input from '~components/Input';
import Button from '~components/Button';
import { URL_FIELD_NAME } from './constants';
import { constructMiniUrlFromSlug, isAppRedirectUrl, isValidUrl } from '~lib/utils/url';
import { getRandomInputUrlMessage, getRandomOutputUrlMessage } from '~lib/utils/messages';
import type { FC, MouseEventHandler } from 'react';
import type { MinifierFormData, MinifierFormProps } from './types';
import styles from './MinifierForm.module.scss';

const MinifierForm: FC<MinifierFormProps> = ({ className, isLoading = false, onReset, onSubmit, slug }) => {
  const [didCopy, setDidCopy] = useState(false);
  const [formMessage, setFormMessage] = useState('');
  const [isAppUrl, setIsAppUrl] = useState(false);

  const {
    register,
    watch,
    setValue,
    handleSubmit,
    formState: { errors },
  } = useForm<MinifierFormData>({ mode: 'onChange' });

  const shouldDisableSubmit = useMemo(() => !isEmpty(errors) || isAppUrl || isLoading, [errors, isAppUrl, isLoading]);

  useEffect(() => {
    const subscription = watch((value) => {
      if (slug) {
        setFormMessage(getRandomOutputUrlMessage());
        return;
      }

      const urlValue = value[URL_FIELD_NAME];

      const isAppUrl = slug === null && isAppRedirectUrl(urlValue ?? '');
      setIsAppUrl(isAppUrl);

      if (isAppUrl) {
        setFormMessage('that url is already shortened. shortening it further would rip the spacetime continuum!');
      } else {
        setFormMessage(isEmpty(urlValue) ? '' : getRandomInputUrlMessage(value[URL_FIELD_NAME] ?? ''));
      }
    });

    return () => subscription.unsubscribe();
  }, [errors, slug, watch]);

  useEffect(() => {
    setValue(URL_FIELD_NAME, slug === null ? '' : constructMiniUrlFromSlug(slug));
  }, [slug, setValue]);

  const handleReset = useCallback(() => {
    setDidCopy(false);
    onReset();
  }, [onReset]);

  const innerSubmitHandler = useCallback(
    (data: MinifierFormData) => {
      slug === null ? onSubmit(data) : handleReset();
    },
    [slug, onSubmit, handleReset],
  );

  const copyToClipboard: MouseEventHandler<HTMLButtonElement> = useCallback(async () => {
    await navigator.clipboard.writeText(constructMiniUrlFromSlug(slug!));
    setDidCopy(true);
  }, [slug]);

  return (
    <form className={cn(className, 'relative')} onSubmit={handleSubmit(innerSubmitHandler)}>
      <div className="relative">
        <Input
          className={styles.urlInput}
          type="url"
          placeholder="paste a url"
          readOnly={slug !== null}
          disabled={isLoading}
          {...register(URL_FIELD_NAME, {
            required: true,
            validate: isValidUrl,
          })}
        />

        {slug !== null && (
          <Button
            type="button"
            title="Copy URL"
            onClick={copyToClipboard}
            className={cn(styles.copyButton, { [styles.copied]: didCopy })}>
            {didCopy ? 'copied!' : 'copy'}
          </Button>
        )}
      </div>

      <Button
        role="submit"
        className={styles.submitButton}
        round
        variant={slug === null ? 'accent' : 'primary'}
        title={slug === null ? 'Submit' : 'Reset'}
        disabled={shouldDisableSubmit}>
        {slug === null ? (
          <ArrowRightIcon className={styles.submitButtonIcon} />
        ) : (
          <RewindIcon className={styles.submitButtonIcon} />
        )}
      </Button>

      <div className="text-slate-600 dark:text-slate-300 text-sm text-center mt-4 h-5">
        {!isEmpty(errors[URL_FIELD_NAME]) ? 'invalid url' : formMessage}
      </div>
    </form>
  );
};

export default MinifierForm;
