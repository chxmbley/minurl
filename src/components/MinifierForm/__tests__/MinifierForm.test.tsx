import { act, render, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import MinifierForm from '~components/MinifierForm';
import * as urlUtils from '~lib/utils/url';
import { constructMiniUrlFromSlug } from '~lib/utils/url';
import { expect } from '@jest/globals';
import { APP_BASE_URL, OUTPUT_URL_MESSAGES } from '~lib/constants';
import { URL_FIELD_NAME } from '~components/MinifierForm/constants';

test('renders a MinifierForm component', () => {
  const { container } = render(<MinifierForm onReset={jest.fn()} onSubmit={jest.fn()} slug={null} />);
  expect(container).toMatchSnapshot();
});

test('calls "onSubmit" when the form is submitted', async () => {
  const inputUrl = 'https://google.com';
  const mockOnSubmit = jest.fn();
  const mockOnReset = jest.fn();

  render(<MinifierForm onReset={mockOnReset} onSubmit={mockOnSubmit} slug={null} />);

  const input = screen.getByPlaceholderText('paste a url');
  const submit = screen.getByTestId('minifier-form-submit');
  const message = screen.getByTestId('minifier-form-message');

  expect(input).toBeInTheDocument();
  expect(input).not.toBeDisabled();
  expect(input).not.toHaveAttribute('readonly');

  expect(submit).toBeInTheDocument();
  expect(submit).toBeDisabled();
  expect(submit).toHaveAttribute('title', 'Submit');

  expect(message.textContent).toBe('');

  await act(async () => {
    await userEvent.type(input, inputUrl);
  });

  expect(input).toHaveValue(inputUrl);
  expect(submit).not.toBeDisabled();

  await act(async () => {
    await userEvent.click(submit);
  });

  expect(mockOnSubmit).toHaveBeenCalledTimes(1);
  expect(mockOnSubmit).toHaveBeenCalledWith({ [URL_FIELD_NAME]: inputUrl });
  expect(mockOnReset).not.toHaveBeenCalled();
});

test('calls "onReset" when the form is submitted when a slug is present', async () => {
  const slug = 'mockslug';
  const mockMiniUrl = 'http://localhost/abc123';
  const constructMiniUrlFromSlugSpy = jest.spyOn(urlUtils, 'constructMiniUrlFromSlug').mockReturnValue(mockMiniUrl);

  const mockOnSubmit = jest.fn();
  const mockOnReset = jest.fn();

  render(<MinifierForm onReset={mockOnReset} onSubmit={mockOnSubmit} slug={slug} />);

  const input = screen.getByPlaceholderText('paste a url');
  const submit = screen.getByTestId('minifier-form-submit');
  const copy = screen.getByTestId('minifier-form-copy');
  const message = screen.getByTestId('minifier-form-message');

  expect(input).toBeInTheDocument();
  expect(input).not.toBeDisabled();
  expect(input).toHaveAttribute('readonly');
  expect(input).toHaveValue(mockMiniUrl);

  expect(copy).toBeInTheDocument();

  expect(submit).toBeInTheDocument();
  expect(submit).not.toBeDisabled();
  expect(submit).toHaveAttribute('title', 'Reset');

  expect(OUTPUT_URL_MESSAGES).toContain(message.textContent);
  expect(constructMiniUrlFromSlugSpy).toHaveBeenCalledTimes(1);

  await act(async () => {
    await userEvent.click(submit);
  });

  expect(mockOnReset).toHaveBeenCalledTimes(1);
  expect(mockOnSubmit).not.toHaveBeenCalled();
  expect(message.textContent).toBe('');
});

test('prevents submit when URL is invalid', async () => {
  const inputUrl = 'https:google.com';
  const mockOnSubmit = jest.fn();
  const mockOnReset = jest.fn();

  render(<MinifierForm onReset={mockOnReset} onSubmit={mockOnSubmit} slug={null} />);

  const input = screen.getByPlaceholderText('paste a url');
  const submit = screen.getByTestId('minifier-form-submit');

  await act(async () => {
    await userEvent.type(input, inputUrl);
    await userEvent.click(submit);
  });

  expect(mockOnSubmit).not.toHaveBeenCalled();
  expect(mockOnReset).not.toHaveBeenCalled();
});

test('prevents submit when URL is formatted as a shortened URL', async () => {
  const inputUrl = `${APP_BASE_URL}/abc123`;
  const mockOnSubmit = jest.fn();
  const mockOnReset = jest.fn();

  render(<MinifierForm onReset={mockOnReset} onSubmit={mockOnSubmit} slug={null} />);

  const input = screen.getByPlaceholderText('paste a url');
  const submit = screen.getByTestId('minifier-form-submit');

  await act(async () => {
    await userEvent.type(input, inputUrl);
    await userEvent.click(submit);
  });

  expect(mockOnSubmit).not.toHaveBeenCalled();
  expect(mockOnReset).not.toHaveBeenCalled();
});
