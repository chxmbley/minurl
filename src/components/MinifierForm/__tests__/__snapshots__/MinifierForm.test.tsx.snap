// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders a MinifierForm component 1`] = `
<div>
  <form
    class="relative"
  >
    <div
      class="relative"
    >
      <input
        class="urlInput input"
        name="url"
        placeholder="paste a url"
        type="url"
      />
    </div>
    <button
      class="submitButton button accent round"
      data-testid="minifier-form-submit"
      disabled=""
      role="submit"
      title="Submit"
    >
      <svg
        aria-hidden="true"
        class="submitButtonIcon"
        fill="currentColor"
        viewBox="0 0 20 20"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          clip-rule="evenodd"
          d="M10.293 3.293a1 1 0 011.414 0l6 6a1 1 0 010 1.414l-6 6a1 1 0 01-1.414-1.414L14.586 11H3a1 1 0 110-2h11.586l-4.293-4.293a1 1 0 010-1.414z"
          fill-rule="evenodd"
        />
      </svg>
    </button>
    <span
      class="block text-slate-600 dark:text-slate-300 text-sm text-center mt-4 h-5"
      data-testid="minifier-form-message"
    />
  </form>
</div>
`;
